### DISCUSSION TITLE
Clarify difference between CLI and Installer

### DISCUSSION THREAD

**Discussion by Kleptine** [openupm/openupm#1382] [Ideas]
This is obviously completely up to the maintainers, but man I really dislike having to download and install an entire separate ecosystem of the programming world, just to install a single Unity package. I get that NodeJS is great, but really this is just a simple command line app, and either Go or Rust would be much less messy.

Just my two cents. I like the project, but it's a hard sell for me unless I was also doing web stuff.

[a comment by an automated account is not shown]

**Comment by favoyang** [comment]
Hi Kleptine,

Thanks for sharing your opinion. I agree that with go or rust, basically, anything can compile to a bite-size binary would be more comfortable for delivering a command-line utility. Install anything like nodejs, python just for this tool is a bit overwhelmed.

While, if you just want to install a package, you can also try the Needle's installer, which is an experimental feature next to the command-line install option on the website.

![image](https://user-images.githubusercontent.com/125390/92777771-1984bd00-f3d3-11ea-80a1-7a746f57c6b7.png)

Or we can wait, until Unity provides a deep link feature for package manager or better integration for 3rd party scoped registry.

I want to point out, because UPM itself is strongly inspired by NPM, and re-used the npm registry server. Lots of toolchain are available here to simplify the CLI implementation. That's a strong benefit when talks to an npm server. And the major reason I choose nodejs/npm over other chooses.

Because of the limited resources and the fact that a large part of our users are asking a GUI integration, the project gonna stay with nodejs/npm.

**Comment by favoyang** [comment]
I'm gonna close this issue, but feel free to continue the discussion.

**Comment by Kleptine** [comment]
Yeah, no worries. Makes sense for convenience. 

I talked with Herbst and it was absolutely unclear that the 'installer.unitypackage' from that link was actually going to modify the package.json. I thought it was just a .unitypackage version of the package itself. Maybe that could be more clear?

**Comment by favoyang** [comment]
Do you have any suggestions for the text?

![image](https://user-images.githubusercontent.com/125390/92843796-b701e000-f417-11ea-91e7-4764dab92d68.png)

**Comment by Kleptine** [comment]
So for me, I was skimming, since all I wanted to do was "just install the
package".

I looked at the 'Download installer.unitypackage' because it's highlighted
so strongly, thought to myself, "Ah, this lets you download it like an
asset!", and then thought nothing more of it, because I didn't want to do
that.

So I'd probably find a way to embed the 'helper package' language into the
main call to action. Something like "Install via Unity" or "Install via
Unity helper".

It occurs to me that the 'download' language was really confusing. I
associate that heavily with 'downloading a specific version of a package,
standalone'. When I think of a package manager, I think of 'adding' or
'installing' a package, not downloading.

**Comment by favoyang** [comment]
How about change the main CTA to
- "Use installer.unitypackage"
- "Use helper.unitypackage"

"Use" hints that you're not going to install the `unitypackage` itself, but use the helper to help you install a upm package.

**Comment by hybridherbst** [comment]
@favoyang can you keep this issue open for the time being & change the title to "Clarify difference between CLI and Installer"? I'll propose some ideas here.

**Comment by hybridherbst** [comment]
OK, so my thinking for this would be that

- installing the CLI requires nodejs as dependency, which might be ok for people who already have that installed but not for people who just want to use some packages
- I think the main confusion here is naming; to me "adding a package to a project" was identical to "installing it", so an "Installer" doing that felt right, but I see the point that this can be confused with "Installer adds all the code directly".
-  clearer and more in line with how Unity names these things in PackMan would be to say "to add the package to your project"

- Possible wordings (very much open for suggestions):
    - "Add the package to your project with setup.unitypackage"
    - "Add the package to your project manifest with installer.unitypackage"


- Additionally, I think the "Installation" box should also include manual instructions / copy-paste helpers, I can help with wording that and setting it up (this is the option for people who don't like the "magic" of the setup.unitypackage, also don't like installing a CLI, but still want to use the package)

@Kleptine @favoyang @marwie further ideas? Otherwise I'll try to make a PR with wording changes in a number of places ("Installation Box", "Package Installer Explanation", and ultimately the Guide should be split in sections for "I want to use the CLI" and "I do not want to use the CLI").

**Comment by favoyang** [comment]
To make it clear, I would like to make the word consistent for our discussion
- installation box
- CLI popup
- installer popup

The **installation box** means,
![image](https://user-images.githubusercontent.com/125390/93012248-4bfc0900-f5d1-11ea-83d7-31c071945032.png)

the **CLI popup** means,
![image](https://user-images.githubusercontent.com/125390/93012257-66ce7d80-f5d1-11ea-8471-2b843e8b6665.png)

the **installer popup** means
![image](https://user-images.githubusercontent.com/125390/93012268-78178a00-f5d1-11ea-94e9-2ef509c75258.png)

**Comment by favoyang** [comment]
If you just want to tweak the text, you can also use the web developer tool to tweak the text (F12 for Chorme), then paste a screenshot like below.

![image](https://user-images.githubusercontent.com/125390/93012426-e446bd80-f5d2-11ea-8f40-3c9289d49f3a.png)

While If you think PR is the best way to go, it's fine as well.

Just 2 cents that a CTA (the text on the button) is meant to be short to get the point.

**Comment by Kleptine** [comment]
I think any of the variations listed here are better than the current text! 

I will say that this depends on the reader seeing 'setup.unitypackage' and realizing that it's a special thing. I could definitely see myself reading that as simply a generic name for all unitypackage's downloaded from this website. 

"installer.unitypackage" is a little better because installer is a more specific word than setup.

[a comment by an automated account is not shown]
